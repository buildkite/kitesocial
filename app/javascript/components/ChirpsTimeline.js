import * as React from 'react';

import consumer from '../channels/consumer';

import Chirp from './Chirp';

class ChirpsTimeline extends React.PureComponent {
  state = {
    chirps: []
  };

  get chirps() {
    return [
      ...this.state.chirps,
      ...this.props.chirps
    ]
  }

  componentDidMount() {
    if (this.props.subscription) {
      this.channel = consumer.subscriptions.create(
        this.props.subscription,
        { received: this.handleChirpReceived }
      );
    }
  }

  componentWillUnmount() {
    if (this.channel) {
      this.channel.unsubscribe();
      this.channel = null;
    }
  }

  handleChirpReceived = (chirp) => {
    this.setState({ chirps: [...this.state.chirps, chirp] })
  }

  render() {
    return (
      <section>
        {this.chirps.map((chirp) => (
          <Chirp key={chirp.id} chirp={chirp} />
        ))}
      </section>
    );
  }
};

export default ChirpsTimeline;
