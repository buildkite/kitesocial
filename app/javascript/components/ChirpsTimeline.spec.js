/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import ChirpsTimeline from './ChirpsTimeline';
import consumer from '../channels/consumer';

// Mock out the chirps so we just check the basic function of the timeline component
jest.mock('./Chirp', () => 'Chirp');

it('renders chirps which are passed in via props', () => {
  const component = renderer.create(
    <ChirpsTimeline
      chirps={[{
        id: 1,
        author: { id: 1,
          name: "alice",
          url: "/users/1" },
        content: "hey @bob @bob @carol @bob @bob",
        created_at: "2021-04-22T01:08:10Z",
        mentions: [
          { id: 2,
            name: "bob",
            url: "/users/2" },
          { id: 3,
            name: "carol",
            url: "/users/3" }
        ],
        updated_at: "2021-04-22T01:08:10Z",
        likes_count: 0,
        liked: false,
        like_url: "/chirps/1/like"
      }]}
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('ActionCable subscriptions', () => {
  it('renders chirps which are passed in via the subscribed channel', () => {
    jest.mock('../channels/consumer');

    const component = renderer.create(
      <ChirpsTimeline
        subscription="MockedChannel"
        chirps={[]}
      />
    );

    const treeBefore = component.toJSON();
    expect(treeBefore).toMatchSnapshot();

    expect(consumer.subscriptions.subscriptions.length).toEqual(1);

    const receivedCallback = consumer.subscriptions.subscriptions[0].received;

    receivedCallback({
      id: 2,
      author: { id: 1,
        name: "alice",
        url: "/users/1" },
      content: "hey again @bob",
      created_at: "2021-04-22T01:08:10Z",
      mentions: [
        { id: 2,
          name: "bob",
          url: "/users/2" }
      ],
      updated_at: "2021-04-22T01:08:10Z",
      likes_count: 0,
      liked: false,
      like_url: "/chirps/1/like"
    });

    const treeAfter = component.toJSON();
    expect(treeAfter).toMatchSnapshot();
  });
});
