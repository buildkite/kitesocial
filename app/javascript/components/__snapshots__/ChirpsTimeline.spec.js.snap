// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionCable subscriptions renders chirps which are passed in via the subscribed channel 1`] = `<section />`;

exports[`ActionCable subscriptions renders chirps which are passed in via the subscribed channel 2`] = `
<section>
  <Chirp
    chirp={
      Object {
        "author": Object {
          "id": 1,
          "name": "alice",
          "url": "/users/1",
        },
        "content": "hey again @bob",
        "created_at": "2021-04-22T01:08:10Z",
        "id": 2,
        "mentions": Array [
          Object {
            "id": 2,
            "name": "bob",
            "url": "/users/2",
          },
        ],
        "updated_at": "2021-04-22T01:08:10Z",
      }
    }
  />
</section>
`;

exports[`renders chirps which are passed in via props 1`] = `
<section>
  <Chirp
    chirp={
      Object {
        "author": Object {
          "id": 1,
          "name": "alice",
          "url": "/users/1",
        },
        "content": "hey @bob @bob @carol @bob @bob",
        "created_at": "2021-04-22T01:08:10Z",
        "id": 1,
        "mentions": Array [
          Object {
            "id": 2,
            "name": "bob",
            "url": "/users/2",
          },
          Object {
            "id": 3,
            "name": "carol",
            "url": "/users/3",
          },
        ],
        "updated_at": "2021-04-22T01:08:10Z",
      }
    }
  />
</section>
`;
